src/kernels/MIOpenFractionalMaxPool.cpp

```diff
@@ -0,0 +1,345 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef MIOPEN_DONT_USE_HIP_RUNTIME_HEADERS
+#include <hip/hip_fp16.h>
+#include <hip/hip_runtime.h>
+#endif
+
+#include "float_types.h"
+#include "hip_atomic.hpp"
+#include "tensor_view.hpp"
+
+__device__ inline int64_t get_interval(
+    FLOAT_ACCUM sample, int64_t index, int64_t input_size, int64_t output_size, int64_t pool_size)
+{
+    if(index == output_size - 1)
+    {
+        return input_size - pool_size;
+    }
+    else
+    {
+        FLOAT_ACCUM alpha = static_cast<FLOAT_ACCUM>(input_size - pool_size) /
+                            static_cast<FLOAT_ACCUM>(output_size - 1);
+        return static_cast<int64_t>((index + sample) * alpha) -
+               static_cast<int64_t>(sample * alpha);
+    }
+}
+
+template <typename T, typename Ti>
+__device__ void fractionalMaxPool2dForward(const T* input,
+                                           T* output,
+                                           Ti* indices,
+                                           T* random_sample,
+                                           int64_t KH,
+                                           int64_t KW,
+                                           tensor_view_t<4> input_tv,
+                                           tensor_view_t<4> output_tv,
+                                           tensor_view_t<4> indices_tv,
+                                           tensor_view_t<3> random_sample_tv)
+{
+    uint64_t gid = blockIdx.x * blockDim.x + threadIdx.x;
+
+    tensor_layout_t<4> layout(output_tv, gid);
+    int64_t n, c, oh, ow;
+    n  = layout.layout[0];
+    c  = layout.layout[1];
+    oh = layout.layout[2];
+    ow = layout.layout[3];
+
+    if(n >= output_tv.size[0])
+        return;
+
+    int64_t pool_h = get_interval(
+        CVT_FLOAT2ACCUM(random_sample[random_sample_tv.get_tensor_view_idx({n, c, 0})]),
+        oh,
+        input_tv.size[2],
+        output_tv.size[2],
+        KH);
+    int64_t pool_w = get_interval(
+        CVT_FLOAT2ACCUM(random_sample[random_sample_tv.get_tensor_view_idx({n, c, 1})]),
+        ow,
+        input_tv.size[3],
+        output_tv.size[3],
+        KW);
+
```

Comment:
Reading modnn source code, it should be similar to following code. Am i missing something here? I suggest to add some test cases with KH != KW
```suggestion
    int64_t pool_h = get_interval(
        CVT_FLOAT2ACCUM(random_sample[random_sample_tv.get_tensor_view_idx({n, c, 1})]),
        oh,
        input_tv.size[2],
        output_tv.size[2],
        KH);
    int64_t pool_w = get_interval(
        CVT_FLOAT2ACCUM(random_sample[random_sample_tv.get_tensor_view_idx({n, c, 0})]),
        ow,
        input_tv.size[3],
        output_tv.size[3],
        KW);

```

Reply:
This might be a bug in Modnn. I couldn't find any other implementation of this operation. There is no difference when changing KH != KW because of the minimal impact of get_interval. By the way, FractionalMaxPool3d has an implementation that uses this order, so let's leave it as it is.